federation/scim/src/main/java/org/keycloak/federation/scim/core/service/UserScimService.java

```diff
@@ -0,0 +1,130 @@
+import org.apache.commons.lang3.BooleanUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.jboss.logging.Logger;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RoleMapperModel;
+import org.keycloak.models.RoleModel;
+import org.keycloak.models.UserModel;
+import org.keycloak.federation.scim.core.ScrimEndPointConfiguration;
+import org.keycloak.federation.scim.core.exceptions.InconsistentScimMappingException;
+import org.keycloak.federation.scim.core.exceptions.SkipOrStopStrategy;
+import org.keycloak.federation.scim.core.exceptions.UnexpectedScimDataException;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+public class UserScimService extends AbstractScimService<UserModel, User> {
+    private static final Logger LOGGER = Logger.getLogger(UserScimService.class);
+
+    public UserScimService(KeycloakSession keycloakSession, ScrimEndPointConfiguration scimProviderConfiguration,
+            SkipOrStopStrategy skipOrStopStrategy) {
+        super(keycloakSession, scimProviderConfiguration, ScimResourceType.USER, skipOrStopStrategy);
+    }
+
+    @Override
+    protected Stream<UserModel> getResourceStream() {
+        return getKeycloakDao().getUsersStream();
+    }
+
+    @Override
+    protected boolean entityExists(KeycloakId keycloakId) {
+        return getKeycloakDao().userExists(keycloakId);
+    }
+
+    @Override
+    protected Optional<KeycloakId> matchKeycloakMappingByScimProperties(User resource) throws InconsistentScimMappingException {
+        Optional<KeycloakId> matchedByUsername = resource.getUserName().map(getKeycloakDao()::getUserByUsername)
+                .map(this::getId);
+        Optional<KeycloakId> matchedByEmail = resource.getEmails().stream().findFirst().flatMap(MultiComplexNode::getValue)
+                .map(getKeycloakDao()::getUserByEmail).map(this::getId);
+        if (matchedByUsername.isPresent() && matchedByEmail.isPresent() && !matchedByUsername.equals(matchedByEmail)) {
+            String inconstencyErrorMessage = "Found 2 possible users for remote user " + matchedByUsername.get() + " - "
+                    + matchedByEmail.get();
+            LOGGER.warn(inconstencyErrorMessage);
+            throw new InconsistentScimMappingException(inconstencyErrorMessage);
+        }
+        if (matchedByUsername.isPresent()) {
+            return matchedByUsername;
+        }
+        return matchedByEmail;
+    }
+
+    @Override
+    protected KeycloakId createEntity(User resource) throws UnexpectedScimDataException {
+        String username = resource.getUserName().filter(StringUtils::isNotBlank)
+                .orElseThrow(() -> new UnexpectedScimDataException(
+                        "Remote Scim user has empty username, can't create. Resource id = %s".formatted(resource.getId())));
+        UserModel user = getKeycloakDao().addUser(username);
```

Comment:
We also need to set the `federationLink` to the UserModel to bind the local user with this provider.